llvm/lib/Target/X86/X86TargetTransformInfo.cpp

```diff
@@ -1550,6 +1550,12 @@ InstructionCost X86TTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
   if (Kind == TTI::SK_InsertSubvector && LT.second.isVector()) {
     int NumElts = LT.second.getVectorNumElements();
     std::pair<InstructionCost, MVT> SubLT = getTypeLegalizationCost(SubTp);
+
+    // If we're inserting into the same legalized type then its not an
+    // insert anymore, its a copy.
+    if (LT.second == SubLT.second)
+      return TTI::TCC_Free;
```

Comment:
This is probably not quite correct. shufflevector <4 x i32>, <4 xi32> , <0,1,2,4> is a subvector insert of 2 vectors with the same types, but it is not free.

Reply:
But wouldn't the SubTp not match?

Reply:
Sorry, just realized what you meant - I'll take another look soon.